### PR TITLE
Operator, char & string literal, separator DFAs

### PR DESCRIPTION
**Pull request by btruhand**

### CODE REVIEW

File: include/identifierDfa.h [hunk 1]
```diff
@@ -0,0 +1,13 @@
+#ifndef __IDENTIFIERDFA_H__
+#define __IDENTIFIERDFA_H__
+
+#include "dfa.h"
+
+class IdentifierDfa : public Dfa {
+  private:
```
Comment (ericluii) [inline include/identifierDfa.h]: indentation -> 4 spaces

File: include/states.h [hunk 2]
```diff
@@ -15,10 +15,31 @@ enum COMPILER_PHASE {
 };
 
 enum DFA_STATES {
+    // Valid number DFA
```
Comment (ericluii) [inline include/states.h]: These are actually shared dfa states, not just for valid number

File: src/identifierDfa.cpp [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    if(('a' <= c && c <= 'z') || ('A' <= c && c <= 'Z') || c == '_' || c == '$') {
+        return DS_ACCEPT;
+    }
+
+    return DS_ERROR;
+}
+
+int isJavaLetterOrDigit(char c) {
+    if(('a' <= c && c <= 'z') || ('A' <= c && c <= 'Z') || ('0' <= c && c <= '9') 
+            || c == '_' || c == '$') {
```
Comment (ericluii) [inline src/identifierDfa.cpp]: ('a' <= c && c <= 'z') || ('A' <= c && c <= 'Z') || ('0' <= c && c <= '9') ||
c == '_' || c == '$')

indentation and stylinggs

File: src/validNumberDfa.cpp [hunk 4]
```diff
@@ -1,7 +1,7 @@
 #include "states.h"
 #include "validNumberDfa.h"
 
-int error(char c) {
+int validNumberError(char c) {
```
Comment (ericluii) [inline src/validNumberDfa.cpp]: If my PR gets merged first, I actually moved error into the base class, so we can all just use that instead.

File: src/validNumberDfa.cpp [hunk 5]
```diff
@@ -10,21 +10,29 @@ int isNonZeroDigit(char c) {
...
 
+    return DS_ZERO;
+}
+
+int noMoreChars(char c) {
     return DS_ERROR;
 }
 
-int isDigit(char c) {
+int isDigitOrUnderscore(char c) {
```
Comment (ericluii) [inline src/validNumberDfa.cpp]: Can I get an example of a valid number that has an underscore in it? o.o

Comment (btruhand) [inline src/validNumberDfa.cpp]: If you go to the Java SE 8 language specification and go to numeric literals you will see some examples that have underscores. 1_2_3 is a valid example and so is 1___2. The language specification outlines more about it

File: src/validNumberDfa.cpp (hunk 4 above)
Comment (btruhand) [inline src/validNumberDfa.cpp]: Nevermind I just saw the change you made to the definition of TransitionFunction and what error looks like

File: include/charStringLiteralDfa.h [hunk 6]
```diff
@@ -0,0 +1,16 @@
+#include "dfa.h"
+#include <string>
+
+#define octalLimit 3
```
Comment (ericluii) [inline include/charStringLiteralDfa.h]: #define OCTAL_LIMIT 3

typical convention for macros

File: include/charStringLiteralDfa.h [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+
+class CharStringLiteralDfa : public Dfa {
```
Comment (ericluii) [inline include/charStringLiteralDfa.h]: I'm starting to make al these final classes :3 so if you could make the change for the new ones thanks :D 

class CharStringLiteralDfa final : public Dfa {

File: include/charStringLiteralDfa.h [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+  private:
+    void initDfa();
+    
+  public:
```
Comment (ericluii) [inline include/charStringLiteralDfa.h]: indentation

File: include/identifierDfa.h [hunk 9]
```diff
@@ -0,0 +1,13 @@
+#ifndef __IDENTIFIERDFA_H__
+#define __IDENTIFIERDFA_H__
+
+#include "dfa.h"
+
+class IdentifierDfa : public Dfa {
```
Comment (ericluii) [inline include/identifierDfa.h]: final class again :3

File: include/operatorDfa.h [hunk 10]
```diff
@@ -0,0 +1,22 @@
+#include "dfa.h"
+#include <string>
+
+class OperatorDfa : public Dfa {
```
Comment (ericluii) [inline include/operatorDfa.h]: final classes :3

File: include/separatorDfa.h [hunk 11]
```diff
@@ -0,0 +1,11 @@
+#include "dfa.h"
+#include <string>
+
+class SeparatorDfa : public Dfa {
```
Comment (ericluii) [inline include/separatorDfa.h]: final (:

**Comment by ericluii** [comment]
Other than my nit-picks LGTM (: good jobs

**Comment by ericluii** [comment]
as discussed offline - will merge as soon as new token types are added, such that I can create a state to token mapping easily, without having to guess what tokens should exist.

**Pull request merged**